src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
-            dd->incRef(tmp0);
+            auto tmp0 = dd->conjugateTranspose(dd_op);
+            auto tmp1 = dd->multiply(reinterpret_cast<dEdge&>(density_root_edge), reinterpret_cast<dEdge&>(tmp0), 0, false);
+            auto tmp2 = dd->multiply(reinterpret_cast<dEdge&>(dd_op), reinterpret_cast<dEdge&>(tmp1), 0, use_density_matrix_type);
+            dd->incRef(tmp2);
+            dEdge::alignDensityEdge(&density_root_edge);
             dd->decRef(density_root_edge);
-            density_root_edge = tmp0;
+            density_root_edge = tmp2;
 
-            if (noiseApplicationWithKrausMatrices) { // was `stochastic_runs == -2`
-                [[maybe_unused]] auto cache_size_before = dd->cn.cacheCount();
+            if (use_density_matrix_type) {
+                dEdge::setDensityMatrixTrue(&density_root_edge);
+            }
 
-                auto seq_targets = targets;
-                for (auto const& control: controls) {
-                    seq_targets.push_back(control.qubit);
+            if (noiseProb > 0 || ampDampingProb > 0) {
+                std::vector used_qubits = targets;
+                for (auto control: controls) {
+                    used_qubits.push_back(control.qubit);
                 }
-                applyDetNoiseSequential(seq_targets);
 
-                [[maybe_unused]] auto cache_size_after = dd->cn.cacheCount();
-                assert(cache_size_after == cache_size_before);
-            } else {
-                signed char maxDepth       = targets[0];
-                auto        control_qubits = op->getControls();
-                for (auto& control: control_qubits) {
-                    if (control.qubit < maxDepth) {
-                        maxDepth = control.qubit;
+                if (sequentialApplyNoise) {
+                    [[maybe_unused]] auto cache_size_before = dd->cn.cacheCount();
+
+                    applyDetNoiseSequential(used_qubits);
+
+                    [[maybe_unused]] auto cache_size_after = dd->cn.cacheCount();
+                    assert(cache_size_after == cache_size_before);
+                } else {
+                    //todo I only must check array elements <=current_v, for the caching
+                    sort(used_qubits.begin(), used_qubits.end(), std::greater<>());
+
+                    [[maybe_unused]] auto cache_size_before = dd->cn.cacheCount();
+
+                    dd::Edge<dd::dNode> nodeAfterNoise = {};
+                    if (use_density_matrix_type) {
+                        dEdge::applyDmChangesToEdges(&density_root_edge, nullptr);
+                        nodeAfterNoise = applyNoiseEffects(density_root_edge, used_qubits, false);
+                        dEdge::revertDmChangesToEdges(&density_root_edge, nullptr);
+                    } else {
+                        nodeAfterNoise = applyNoiseEffects(density_root_edge, used_qubits, true);
```

Comment:
If the `applyNoiseEffects` function is moved to the DD Package, it would make sense, to encapsulate this as a separate function.

In addition: Can't the use of the density matrix type be detected by checking whether the respective flag is set on the root edge?

Reply:
The entire noise functionality is encapsulated into the QFR. 

No. Depending on where I am within my density matrix decision diagram, I sometimes need to treat density matrix nodes as normal matrix nodes.